### PR TITLE
Task/dat 649 merge authenticator implementations

### PR DESCRIPTION
**Pull request by hb0**
This PR moves `libs/api` and `libs/model` to the Collector to be built as public libraries and to be used by all our Vert.X APIs.

This reduces the code to be maintained and for the Vert.X APIs to be compatible, e.g. the data-collector and the data-exporter/provider. This way they can all use the same user-database and the same authenticators.

The same structure was used as in our backend to keep the number of `build.gradle` variants to be maintained small.

To use our default deployment workflow, this repository now only publishes the Jars to the Github Package Registry.
The Docker Images will be published by a separate repository to the Github Package Registry instead of the Docker Registry as before.

### CODE REVIEW

**Review by hb0 — COMMENTED** [review]
Added comments for discussion

File: libs/api/src/main/java/de/cyface/api/Hasher.java [hunk 1]
```diff
@@ -46,7 +65,7 @@ public Hasher(final HashingStrategy hashingStrategy, final byte[] salt) {
      * @return The hashed and salted password
      */
     public String hash(final String password) {
-        return hashingStrategy.hash("pbkdf2",
+        return hashingStrategy.hash(HashAlgorithm.PBKDF2.name().toLowerCase(), // FIXME: stronger hashing algorithm recommended
```
[the remark this answers is not in the document]
Comment (muthenberg) [reply libs/api/src/main/java/de/cyface/api/Hasher.java:68]: ok

File: libs/build.gradle [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  dependencies {
+    implementation "org.slf4j:slf4j-api:$slf4jVersion"
+    implementation "org.apache.commons:commons-lang3:$commonsLangVersion"
+  }
+
+  // The following needs to be wrapped in afterEvaluate. I have no idea why, but the solution is described here: https://github.com/akhikhl/gretty/issues/306
+  project.afterEvaluate {
+
+    // This task creates an obfuscated Jar file from the one created by the jar task and puts it into the 'obf' folder of the build directory
+    task createObfuscated(type: ProGuardTask, dependsOn: 'build') {
```
[the remark this answers is not in the document]
Comment (muthenberg) [reply libs/build.gradle:33]: Yeah. It is not only used for obfuscation but it also can speed up things.

File: libs/api/src/main/java/de/cyface/api/ServerConfig.java [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+                    keyBuilder.append(line);
+                }
+            }
+        }
+
+        final var key = keyBuilder.toString();
+        Validate.notNull(key, String.format(
+                "Unable to load key for JWT authentication. Did you provide a valid PEM file using the parameter %s.",
+                keyParameter.key()));
+        return key;
```
[the remark this answers is not in the document]
Comment (muthenberg) [reply libs/api/src/main/java/de/cyface/api/ServerConfig.java:251]: That depends on the format of the keys. If you need the key in between lines starting with ----- you need to use the complex variant.

**Review by hb0 — COMMENTED** [review]
Cannot get the Tests to work with RS256 ...

File: executables/collector/src/test/java/de/cyface/collector/AuthorisationTest.java [hunk 4]
```diff
@@ -64,26 +70,29 @@
...
+                .addPubSecKey(new PubSecKeyOptions()
+                        .setAlgorithm(JWT_HASH_ALGORITHM)
+                        .setBuffer(privateKey)));
+
         final var issuer = "localhost:8080/api/v2/";
+        final var body = new JsonObject().put("username", "test").put("password", "test-password")
+                .put("aud", issuer).put("iss", issuer);
+        final var jwtOptions = new JWTOptions()
+                .setAlgorithm(JWT_HASH_ALGORITHM).setIssuer(issuer).setAudience(Collections.singletonList(issuer));
+        final var token = provider.generateToken(body, jwtOptions);
```
Comment (hb0) [inline executables/collector/src/test/java/de/cyface/collector/AuthorisationTest.java:87]: Before we set the JWTOptions in provider.setJwtOptions, which did not work after changing to RS256.
I got multiple errors when executing the Test.

When I set the JWTOptions in `generateToken()` as suggested in the Vert.X Dokumentation everything works with RS256, too.

File: libs/api/src/main/java/de/cyface/api/ServerConfig.java [hunk 5]
```diff
@@ -229,26 +225,7 @@ public String extractKey(final Vertx vertx, final Parameter keyParameter)
...
-                }
-            }
-        }
-
-        final var key = keyBuilder.toString();
-        Validate.notNull(key, String.format(
-                "Unable to load key for JWT authentication. Did you provide a valid PEM file using the parameter %s.",
-                keyParameter.key()));
-        return key;
+        return Files.readString(Path.of(keyFilePath));
```
Comment (hb0) [inline libs/api/src/main/java/de/cyface/api/ServerConfig.java:228]: With RS256 the `----` lines are needed, or else I get the error "pem files do not contain enough lines".
The Vert.X dokumentation also injects the keys including these lines.

**Review by hb0 — COMMENTED** [review]
Finally got RS256 to work. Was a very hard to find issue, see comments below.

**Review by muthenberg — APPROVED** [review]

File: executables/collector/src/main/docker/docker-compose.yml [hunk 6]
```diff
@@ -55,7 +56,7 @@ services:
   api:
     container_name: collector_api
     build: .
-    image: cyfacede/collector:5.2.1
+    image: data-collector:0.0.0
```
[the remark this answers is not in the document]
Comment (hb0) [reply executables/collector/src/main/docker/docker-compose.yml:59]: Right, new branching model. I created a task for this https://cyface.atlassian.net/browse/CY-5603

**Pull request merged**